docs/architecture.md

```diff
@@ -0,0 +1,185 @@
+
+Once the model class is successfully built, a core schema for the model will be constructed
+and set as the [`__pydantic_core_schema__`][pydantic.main.BaseModel.__pydantic_core_schema__]
+attribute (this is done by the `GenerateSchema` class). A core schema is a structured Python
+dictionary (represented using [`TypedDict`][typing.TypedDict] definitions) representing a
+specific validation and serialization logic. It is the core data structure used to communicate
+between the `pydantic` and `pydantic-core` packages. Every core schema has a required `type` key,
+and extra properties depending on this `type`.
+
+!!! note
+    It is not possible to define a custom core schema. A core schema needs to be understood by the
+    `pydantic-core` package, and as such we only support a fixed number of core schema types.
+    This is also part of the reason why the `GenerateSchema` is private and not documented.
+
+    The core schema definitions can be found in the [`pydantic_core.core_schema`][] module.
+
+To illustrate what a core schema looks like, we will take the example of the
+[`bool`][pydantic_core.core_schema.bool_schema] core schema:
+
+```python lint="skip" test="skip"
+class BoolSchema(TypedDict, total=False):
+    type: Required[Literal['bool']]
+    strict: bool
+    ref: str
+    metadata: Any
+    serialization: SerSchema
+```
+
+When defining a Pydantic model with a boolean field:
+
+```python
+from pydantic import BaseModel, Field
+
+
+class Model(BaseModel):
+    foo: bool = Field(strict=True)
+```
+
+The core schema for the `foo` field will look like:
+
+```python
+{
+    'type': 'bool',
+    'strict': True,
+}
+```
+
+As seen in the [`BoolSchema`][pydantic_core.core_schema.bool_schema] definition,
+the serialization logic is also defined in the core schema.
+If we were to define a custom serialization function for `foo` (1), the `serialization` key would look like:
+{ .annotate }
+
+1.  For example using the [`field_serializer`][pydantic.functional_serializers.field_serializer] decorator.
+
+```python lint="skip" test="skip"
+{
+    'type': 'function-plain',
+    'function': <function Model.serialize_foo at 0x111>,
+    'is_field_serializer': True,
+    'info_arg': False,
+    'return_schema': {'type': 'int'},
+}
+```
+
+Note that this is also a core schema definition, just that it is only relevant for `pydantic-core` during serialization.
+Core schemas cover a broad scope, and are used whenever we want to communicate between the Python and Rust side.
+
+### JSON Schema generation
+
+You may have noticed that the previous serialization core schema has a `return_schema` key.
+This is because the core schema is also used to generate the corresponding JSON Schema.
+
+The JSON Schema generation of a model is handled by the [`GenerateJsonSchema`][pydantic.json_schema.GenerateJsonSchema] class.
+The [`generate`][pydantic.json_schema.GenerateJsonSchema.generate] method is the main entry point and is given the core schema of that model.
```

Comment:
Perhaps an example here could be beneficial - like what would the json schema look like for that `bool` schema you demoed above?